pennylane/transforms/hamiltonian_expand.py

```diff
@@ -106,26 +106,34 @@ def hamiltonian_expand(tape, group=True):
             "Passed tape must end in `qml.expval(H)`, where H is of type `qml.Hamiltonian`"
         )
 
-    if group:
-        hamiltonian.simplify()
-        return qml.transforms.measurement_grouping(tape, hamiltonian.ops, hamiltonian.coeffs)
+    if group or hamiltonian.grouping_indices is not None:
+        # if explicitly requested or already available,
+        # compute the grouping information
+        coeffs_groupings, obs_groupings = hamiltonian.get_groupings()
+    else:
+        # else make each observable its own group
+        obs_groupings = [[obs] for obs in hamiltonian.ops]
+        coeffs_groupings = [hamiltonian.coeffs[i] for i in range(len(hamiltonian.ops))]
 
     # create tapes that measure the Pauli-words in the Hamiltonian
     tapes = []
-    for ob in hamiltonian.ops:
-        # we need to create a new tape here, because
-        # updating metadata of a copied tape is error-prone
-        # when the observables were changed
+    for obs in obs_groupings:
+
         with tape.__class__() as new_tape:
             for op in tape.operations:
-                qml.apply(op)
-            qml.expval(ob)
+                op.queue()
+
+            for o in obs:
+                qml.expval(o)
+
+        new_tape = new_tape.expand(stop_at=lambda obj: True)
         tapes.append(new_tape)
 
-    # create processing function that performs linear recombination
     def processing_fn(res):
         dot_products = [
-            qml.math.dot(qml.math.squeeze(res[i]), hamiltonian.coeffs[i]) for i in range(len(res))
+            # the order is important here, because r
+            # may have an extra dimension if tape was evaluated with a distribution of shots
+            qml.math.dot(qml.math.squeeze(r), c) for c, r in zip(coeffs_groupings, res)
```

Comment:
I had a day's worth of issues with the shapes here, and decided to always squeeze everything - that seems to work